File: 3/3.rb
#! /bin/env ruby 

def is_prime(num)
    return false if num < 2 || num % 2 == 0

    i = 3
    while i < Math.sqrt(num) + 1
        return false if num % i == 0 

        i += 2 
    end 

    true
end 

num = 600851475143
largest_prime = 0

while num % 2 == 0 
    num /= 2
end 

while num % 3 == 0 
    num /= 3 
end 

prime = 5
max = Math.sqrt(num) + 1 
while prime < max 
    while num % prime == 0
        largest_prime = prime 

        num /= prime
    end 

    loop do 
        prime += 6
        break if is_prime(prime)
    end 
end 

p largest_prime 

File: 3/3.go
package main 

import (
	"math"
)

func isPrime(num int) bool {
	if num < 2 || num % 2 == 0 {
		return false 
	}

	for i := 3; i < int(math.Sqrt(float64(num))); i += 2 {
		if num % i == 0 {
			return false 
		}
	}

	return true
}

func main() {
	num := 600851475143
	largestPrime := 0

	for ; num % 2 == 0; {
		num = int(num / 2)
	}

	for ; num % 3 == 0; {
		num = (num / 3)
	}

	prime := 5
	max := int(math.Sqrt(float64(num))) + 1
	for ; prime < max; {
		for ; num % prime == 0; num = int(num / prime) {
			largestPrime = prime 
		}

		for ;; {
			prime += 6

			if isPrime(prime) {
				break 
			}
		}
	}

	println(largestPrime)
}
